### ISSUE TITLE
Enhancement: Abort callbacks in image processing functions

### ISSUE THREAD

**Issue by MrKepzie** [OpenImageIO/oiio#1260]
Currently there seem to be only the write_image that allow an abort callback to be passed, i.e: where we can notify the processing code to stop at any given moment. 
All the rest of the image processing functions do not have the ability to abort. Could this be generalized to the potentially costly image processing functions (such as ImageBufAlgo::resize etc...) ?

**Comment by lgritz** [comment]
Sorry, this one slipped through the cracks.

The idea was that read_image/write_image, for very large images (or very slow disks?) might take long enough that it would be helpful for an interactive app to have a meaningful progress indicator and thus needs a callback. The ability for the callback to also be able to say "abort" via the return code was secondary but came for "free."

I don't know if the callback for read/write_image turned out to be a good idea or is even necessary. I have not been tempted to add it anywhere else. I'd like to think that the IBA functions are fast enough that progress meters are not necessary. I'm cautious about adding extra parameters to dozens of other functions, I think there is a real cost (cognitively, not necessarily performance) to adding API complexity without a good reason.

For many of the IBA functions we face tradeoffs in choice parameters. For almost every function, there COULD have been another 10 parameters to control every last detail. (Even something as simple as add(): should there be a control to allow it to "saturate" or clamp? Should it allow another optional image to provide a "mask" giving more flexible control of where it happens than an ROI? Etc.) At some point, we need to draw a line and keep the API signature as simple as possible while still being applicable to most use cases, and if there is the occasional app that needs slightly different behavior, it can always supply its own function! Being open source with the implementation completely spelled out, the amount of effort is very low for an app to do something similar to an existing IBA but with slightly different controls and behavior. In a closed source library, the pressure would be higher to make a proliferation of functions or parameters to cover every possible use case.

Long explanation, but it's an important part of the API design philosophy.

Or am I totally misunderstanding that there is a compelling use case for having progress/abort callback parameters to absolutely everything?

Would a more general solution be to create a wrapper that launches a thread to run an IBA function (as a lambda?), and then aborting just means killing the thread before it's done? (Can that be done without killing the whole app?)

**Comment by MrKepzie** [comment]
That request was very oriented towards our own needs, we now use another solution. 

The idea is that in our compositing graph, we try to compute as much of the needed portion of an image in one pass to avoid what we call *overlapping tiles*. 

Basically if we were to split up an image into small tiles and process each tiles on different threads, some filters (such as a gaussian blur) would process actually a little bit more than just a tile. All summed up, some portions of the image would be computed several times by different threads.
So at some point we end up with a *big tile* which we give to a node of the graph to process. 

When the user makes a change to any property, the compositing graph needs to abort immediately whatever is going on (that is no longer relevant) and compute a new version of the image.

To be able to be responsive, every node must be able to abort quickly their processing with a callback. 

We had a node using IBA::resize, and when processing large images it was very slow to abort (roughly several seconds) which would be very annoying for the user. 

Basically we wrote our own internal resize functions for our needs.

I understand totally your point, though I'm not sure copying the code around for specific needs is the right way of doing things either(and it's not what we did). 

This request can be closed.

**Comment by lgritz** [comment]
I understand the use case, and I think you did the right thing -- write your own, abortable, version of the one (or few) functions for which this is a true problem. 

I also think that this might mean that resize in particular might be ripe for a look at how to optimize it more. At my studio we use it (and lots of IBA) mostly for batch processing from scripts, rather than in interactive applications, so the practical difference between 100ms and 2s is essentially negligible. Which doesn't mean I don't think speed is important, just that there is a threshold below which nobody down the hall is going to yell at me, so it can go unnoticed if nobody else points it out. :-)
